### ISSUE TITLE
channeldb: migration code should run in separate transactions

### ISSUE THREAD

**Issue by Crypt-iQ** [lightningnetwork/lnd#7296]
If we have an outdated channel database, the channeldb code will attempt to run the migrations to get the database up to date. The problem is that every migration besides _one_ optional migration is run in a single transaction. If the database is older and needs to do several intensive migrations, this can lead to high RAM usage. 

See `syncVersions`: https://github.com/lightningnetwork/lnd/blob/81b17ef4fde9c73d57e699a97778454198f3893d/channeldb/db.go#L1473-L1486

**Comment by yyforyongyu** [comment]
This has come up once when working on that db reduction. The change would be straightforward: we put migrations into multiple transactions, with each transaction capped with a predefined memory limit. Implementation-wise, my draft would be,
- run each migration in a single db tx.
- if the migration exceeds the mem limit, split it into multiple txes.

The most complicated part would be using multiple db txes but still supporting `dry-run` mode, which would require us to write data to disk, and delete them once the dry-run is finished.

**Comment by Crypt-iQ** [comment]
I don't think we can know the memory while it's running without using something like pprof to grab a heap dump, but that seems like some sort of bad programming practice. I think the best we could do here without using pprof is to figure out what operations are memory intensive and then split those operations up.

**Comment by yyforyongyu** [comment]
I think we can use `runtime` to create something like [this](https://gist.github.com/j33ty/79e8b736141be19687f565ea4c6f4226) to get the current mem usage.

**Comment by guggero** [comment]
Shouldn't the individual migrations be independent in the first place? So would moving the loop over the migrations outside of the `kvdb.Update` not in itself alleviate the problem somewhat? And there's always the option of upgrading versions step-by-step (not very user friendly, I know, but we're talking about edge cases with huge databases here, right?).
